Sources/App/routes.swift

```diff
@@ -34,102 +34,107 @@ func routes(_ app: Application) throws {
+
         guard let origin = Environment.get("ORIGIN") else {
             throw Abort(.internalServerError)
         }
-        
-        let credential = try WebAuthn.parseRegisterCredentials(registerData, challengeProvided: challenge, origin: origin, logger: req.logger)
-        
-        guard let username = req.session.data["username"], let userIDString = req.session.data["userID"], let userID = UUID(uuidString: userIDString) else {
+
+        let credential = try req.webAuthn.parseRegisterCredentials(registerData, challengeProvided: challenge, origin: origin, logger: req.logger)
+
+        guard let userIDString = req.session.data["userID"],
+            let userID = UUID(uuidString: userIDString),
+            let user = try await User.find(userID, on: req.db) else {
             throw Abort(.badRequest)
         }
-        
-        let user = User(id: userID, username: username)
-        try await user.save(on: req.db)
-        
+
         let webAuthnCredential = WebAuthnCredential(id: credential.credentialID, publicKey: credential.publicKey.pemRepresentation, userID: userID)
         try await webAuthnCredential.save(on: req.db)
-        
+
         req.auth.login(user)
-        
+
         return .ok
     }
-    
+
     // step 1 for authentication
     authSessionRoutes.get("authenticate") { req -> StartAuthenticateResponse in
         let username = try req.query.get(String.self, at: "username")
         guard let user = try await User.query(on: req.db).filter(\.$username == username).first() else {
             throw Abort(.unauthorized)
         }
-        let challenge = [UInt8].random(count: 32).base64
-        let encodedChallenge = challenge.replacingOccurrences(of: "+", with: "-")
-            .replacingOccurrences(of: "/", with: "_")
-            .replacingOccurrences(of: "=", with: "")
+        let challenge = try req.webAuthn.generateChallenge()
+        let encodedChallenge = challenge.base64URLEncodedString()
         req.logger.debug("Authenticate Challenge is \(encodedChallenge)")
         req.session.data["challenge"] = encodedChallenge
         req.session.data["userID"] = try user.requireID().uuidString
         let credentials = try await user.$credentials.get(on: req.db).map { credential -> WebAuthnCredential in
             // We need to convert the IDs to base64 encoded from base64 URL encoded
-            var id = credential.id!.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
+            var id = String.base64(fromBase64URLEncoded: credential.id!)
             while id.count % 4 != 0 {
                 id = id.appending("=")
             }
             return WebAuthnCredential(id: id, publicKey: credential.publicKey, userID: credential.$user.id)
         }
-        return StartAuthenticateResponse(challenge: challenge, credentials: credentials)
+        return StartAuthenticateResponse(challenge: challenge.base64String(), credentials: credentials)
```

Comment:
I guess we can probably roll these into helpers too

Reply:
yep that endpoint i didn't touch yet